File: serverless/javascript/differential/encryption-header.test.mjs
// Property test for the remote encryption key header of the serverless
// driver, from the shared spec's `tests.encryption_header` entry
// (serverless/conformance/differential/spec/ops.json; described in
// serverless/conformance/differential/operations.md and PROTOCOL.md
// section 3.1): a driver configured with key K attaches
// `x-turso-encryption-key: K` to EVERY HTTP request — pipeline and cursor
// endpoints alike — and a driver configured without a key never sends the
// header.
//
// Runs against a local stub HTTP server that records request headers and
// speaks just enough of the protocol for a statement to complete. It needs
// no Turso Cloud database and runs unconditionally — no environment
// configuration, never skips.

import { createServer } from 'node:http';
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import test from 'ava';
import * as fc from 'fast-check';

import { connect, ENCRYPTION_KEY_HEADER } from '@tursodatabase/serverless';

// ---------------------------------------------------------------------------
// Load spec from ops.json
// ---------------------------------------------------------------------------

const __dirname = dirname(fileURLToPath(import.meta.url));
const _SPEC = JSON.parse(
  readFileSync(
    join(__dirname, '..', '..', 'conformance', 'differential', 'spec', 'ops.json'),
    'utf8'
  )
);

const ENC = _SPEC.tests.encryption_header;

// ---------------------------------------------------------------------------
// Stub Hrana server — records every request's headers and answers with the
// minimal protocol shapes from PROTOCOL.md sections 5–7.
// ---------------------------------------------------------------------------

// One pipeline result per request, in request order (PROTOCOL.md section 5.2).
function pipelineResult(request) {
  switch (request.type) {
    case 'execute':
      return {
        type: 'ok',
        response: {
          type: 'execute',
          result: {
            cols: [{ name: 'x', decltype: '' }],
            rows: [[{ type: 'integer', value: '1' }]],
            affected_row_count: 0,
            last_insert_rowid: null,
          },
        },
      };
    case 'describe':
      return {
        type: 'ok',
        response: {
          type: 'describe',
          result: { params: [], cols: [{ name: 'x', decltype: '' }], is_explain: false, is_readonly: true },
        },
      };
    case 'get_autocommit':
      return { type: 'ok', response: { type: 'get_autocommit', is_autocommit: true } };
    default:
      // sequence, close, batch — an empty ok response suffices.
      return { type: 'ok', response: { type: request.type } };
  }
}

// Newline-separated cursor stream (PROTOCOL.md section 7.2): first the
// cursor response, then step_begin/row/step_end per step. The stub is
// always in autocommit, so every step — including the driver's trailing
// is_autocommit probe — executes.
function cursorBody(request) {
  const lines = [{ baton: 'stub-baton', base_url: null }];
  request.batch.steps.forEach((step, i) => {
    const wantRows = step.stmt.want_rows;
    lines.push({ type: 'step_begin', step: i, cols: wantRows ? [{ name: 'x', decltype: '' }] : [] });
    if (wantRows) {
      lines.push({ type: 'row', row: [{ type: 'integer', value: '1' }] });
    }
    lines.push({ type: 'step_end', affected_row_count: 0, last_insert_rowid: null });
  });
  lines.push({ type: 'replication_index', replication_index: '0' });
  return lines.map((l) => JSON.stringify(l)).join('\n') + '\n';
}

// Starts the stub on 127.0.0.1 and resolves to { port, requests, close }.
// `requests` accumulates { path, headers } for every request received.
function startStubServer() {
  const requests = [];
  const server = createServer((req, res) => {
    let body = '';
    req.on('data', (chunk) => { body += chunk; });
    req.on('end', () => {
      // Node lowercases incoming header names in req.headers.
      requests.push({ path: req.url, headers: { ...req.headers } });
      const request = JSON.parse(body);
      if (req.url === '/v3/pipeline') {
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({
          baton: null,
          base_url: null,
          results: request.requests.map(pipelineResult),
        }));
      } else if (req.url === '/v3/cursor') {
        res.writeHead(200, { 'Content-Type': 'application/json' });
        res.end(cursorBody(request));
      } else {
        res.writeHead(404);
        res.end();
      }
    });
  });
  return new Promise((resolve) => {
    server.listen(0, '127.0.0.1', () => {
      resolve({
        port: server.address().port,
        requests,
        close: () => new Promise((r) => server.close(r)),
      });
    });
  });
}

// ---------------------------------------------------------------------------
// Drive the stub through both endpoints: exec() goes to /v3/pipeline
// (sequence + get_autocommit), all() goes to /v3/cursor, and close() sends
// a final /v3/pipeline close for the baton the cursor issued.
// ---------------------------------------------------------------------------

async function runStatements(config) {
  const conn = connect(config);
  try {
    await conn.exec('SELECT 1');
    await conn.all('SELECT 1');
  } finally {
    await conn.close();
  }
}

function assertEndpointCoverage(requests) {
  const paths = new Set(requests.map((r) => r.path));
  if (requests.length < 3 || !paths.has('/v3/pipeline') || !paths.has('/v3/cursor')) {
    throw new Error(
      `expected pipeline and cursor requests, got: ${JSON.stringify(requests.map((r) => r.path))}`
    );
  }
}

// ---------------------------------------------------------------------------
// The properties
// ---------------------------------------------------------------------------

// Keys drawn from the spec's base64 alphabet, optionally with trailing
// '=' padding.
const arbKey = fc.tuple(
  fc.string({
    minLength: ENC.key_min_len,
    maxLength: ENC.key_max_len,
    unit: fc.constantFrom(...ENC.key_alphabet),
  }),
  fc.constantFrom('', '=', '==')
).map(([body, pad]) => body + pad);

test.serial('encryption header: configured key sent on every request', async (t) => {
  // The driver's exported constant must match the spec's header name.
  t.is(ENCRYPTION_KEY_HEADER, ENC.header);

  const server = await startStubServer();
  try {
    await fc.assert(
      fc.asyncProperty(arbKey, async (key) => {
        server.requests.length = 0;
        await runStatements({
          url: `http://127.0.0.1:${server.port}`,
          authToken: 'x',
          remoteEncryptionKey: key,
        });
        assertEndpointCoverage(server.requests);
        for (const req of server.requests) {
          if (req.headers[ENC.header] !== key) {
            throw new Error(
              `request to ${req.path} carried header ${JSON.stringify(req.headers[ENC.header])}, ` +
              `expected ${JSON.stringify(key)}`
            );
          }
        }
      }),
      { numRuns: ENC.num_examples, verbose: 1 }
    );
    t.pass();
  } catch (e) {
    t.fail(e.message);
  } finally {
    await server.close();
  }
});

test.serial('encryption header: absent on every request without a key', async (t) => {
  const server = await startStubServer();
  try {
    await runStatements({
      url: `http://127.0.0.1:${server.port}`,
      authToken: 'x',
    });
    assertEndpointCoverage(server.requests);
    for (const req of server.requests) {
      t.is(
        req.headers[ENC.header],
        undefined,
        `request to ${req.path} must not carry ${ENC.header}`
      );
    }
    t.pass();
  } finally {
    await server.close();
  }
});

File: serverless/go/encryption_header_test.go
// Property test for the remote encryption key header, implementing the
// `encryption_header` entry of serverless/conformance/differential/spec/ops.json:
// for any key K drawn from the spec's alphabet, a driver configured with K
// sends `x-turso-encryption-key: K` on every HTTP request — pipeline and
// cursor endpoints alike — and a driver with no key never sends the header
// (PROTOCOL.md section 3.1). Runs against a local stub server that records
// request headers; needs no live database and never skips.

package turso

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"net/http/httptest"
	"net/url"
	"sync"
	"testing"
)

// Parameters of the `encryption_header` spec entry.
const (
	encHeaderName     = "x-turso-encryption-key"
	encKeyAlphabet    = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
	encKeyMinLen      = 1
	encKeyMaxLen      = 64
	encHeaderExamples = 20
)

// recordedRequest is one HTTP request seen by the stub server.
type recordedRequest struct {
	path      string
	hasHeader bool
	key       string
}

// stubServer speaks just enough of the SQL over HTTP protocol for a
// statement to complete, recording the encryption key header of every
// request.
type stubServer struct {
	t   *testing.T
	srv *httptest.Server

	mu       sync.Mutex
	requests []recordedRequest
}

func newStubServer(t *testing.T) *stubServer {
	s := &stubServer{t: t}
	s.srv = httptest.NewServer(http.HandlerFunc(s.handle))
	t.Cleanup(s.srv.Close)
	return s
}

func (s *stubServer) handle(w http.ResponseWriter, r *http.Request) {
	values := r.Header.Values(encHeaderName)
	s.mu.Lock()
	s.requests = append(s.requests, recordedRequest{
		path:      r.URL.Path,
		hasHeader: len(values) > 0,
		key:       r.Header.Get(encHeaderName),
	})
	s.mu.Unlock()

	switch r.URL.Path {
	case "/v3/pipeline":
		s.handlePipeline(w, r)
	case "/v3/cursor":
		s.handleCursor(w, r)
	default:
		s.t.Errorf("stub server: unexpected path %q", r.URL.Path)
		http.NotFound(w, r)
	}
}

// handlePipeline answers a pipeline request (PROTOCOL.md section 5) with one
// ok result per request, in request order.
func (s *stubServer) handlePipeline(w http.ResponseWriter, r *http.Request) {
	var preq pipelineRequest
	if err := json.NewDecoder(r.Body).Decode(&preq); err != nil {
		s.t.Errorf("stub server: invalid pipeline request: %v", err)
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	results := make([]any, len(preq.Requests))
	for i, req := range preq.Requests {
		response := map[string]any{"type": req.Type}
		switch req.Type {
		case "get_autocommit":
			response["is_autocommit"] = true
		case "describe":
			response["result"] = map[string]any{"params": []any{}}
		case "execute":
			response["result"] = map[string]any{
				"cols":               []any{},
				"rows":               []any{},
				"affected_row_count": 0,
				"last_insert_rowid":  nil,
			}
		}
		results[i] = map[string]any{"type": "ok", "response": response}
	}
	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(map[string]any{
		"baton":    "stub-baton",
		"base_url": nil,
		"results":  results,
	}); err != nil {
		s.t.Errorf("stub server: failed to encode pipeline response: %v", err)
	}
}

// handleCursor answers a cursor request (PROTOCOL.md section 7) with a
// newline-separated body: the cursor response line, then step_begin, row,
// and step_end entries for every step. All conditions (including the
// trailing autocommit probe the driver appends) are treated as satisfied.
func (s *stubServer) handleCursor(w http.ResponseWriter, r *http.Request) {
	var creq cursorRequest
	if err := json.NewDecoder(r.Body).Decode(&creq); err != nil {
		s.t.Errorf("stub server: invalid cursor request: %v", err)
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	lines := []any{
		map[string]any{"baton": "stub-baton", "base_url": nil},
	}
	for i := range creq.Batch.Steps {
		lines = append(lines,
			map[string]any{
				"type": "step_begin",
				"step": i,
				"cols": []any{map[string]any{"name": "a", "decltype": nil}},
			},
			map[string]any{
				"type": "row",
				"row":  []any{map[string]any{"type": "integer", "value": "1"}},
			},
			map[string]any{
				"type":               "step_end",
				"affected_row_count": 1,
				"last_insert_rowid":  1,
			},
		)
	}
	enc := json.NewEncoder(w)
	for _, line := range lines {
		if err := enc.Encode(line); err != nil {
			s.t.Errorf("stub server: failed to encode cursor entry: %v", err)
			return
		}
	}
}

// randomEncryptionKey draws a key per the spec entry: a length in
// [encKeyMinLen, encKeyMaxLen] over encKeyAlphabet, plus optional trailing
// base64 '=' padding.
func randomEncryptionKey(rng *rand.Rand) string {
	n := encKeyMinLen + rng.Intn(encKeyMaxLen-encKeyMinLen+1)
	key := make([]byte, 0, n+2)
	for i := 0; i < n; i++ {
		key = append(key, encKeyAlphabet[rng.Intn(len(encKeyAlphabet))])
	}
	for i := rng.Intn(3); i > 0; i-- {
		key = append(key, '=')
	}
	return string(key)
}

func TestEncryptionHeaderProperty(t *testing.T) {
	rng := rand.New(rand.NewSource(20260807))
	for i := 0; i < encHeaderExamples; i++ {
		key := randomEncryptionKey(rng)
		viaDSN := i%2 == 0
		t.Run(fmt.Sprintf("key_%02d", i), func(t *testing.T) {
			runEncryptionHeaderExample(t, key, viaDSN)
		})
	}
	t.Run("no_key_dsn", func(t *testing.T) { runEncryptionHeaderExample(t, "", true) })
	t.Run("no_key_connector", func(t *testing.T) { runEncryptionHeaderExample(t, "", false) })
}

// runEncryptionHeaderExample opens a database configured with key (or with
// no key when empty), runs one exec-style statement and one query, and
// asserts every request the stub saw carried (or did not carry) the header.
func runEncryptionHeaderExample(t *testing.T, key string, viaDSN bool) {
	stub := newStubServer(t)

	var db *sql.DB
	if viaDSN {
		dsn := stub.srv.URL
		if key != "" {
			dsn += "?remote_encryption_key=" + url.QueryEscape(key)
		}
		var err error
		db, err = sql.Open("turso-serverless", dsn)
		if err != nil {
			t.Fatal(err)
		}
	} else {
		connector := NewConnector(stub.srv.URL, "")
		if key != "" {
			connector = connector.WithRemoteEncryptionKey(key)
		}
		db = sql.OpenDB(connector)
	}

	// A multi-statement exec runs as a sequence request on /v3/pipeline; a
	// query runs on /v3/cursor.
	if _, err := db.Exec("CREATE TABLE t (a); INSERT INTO t VALUES (1)"); err != nil {
		t.Fatal(err)
	}
	var v int64
	if err := db.QueryRow("SELECT a FROM t").Scan(&v); err != nil {
		t.Fatal(err)
	}
	if v != 1 {
		t.Fatalf("got %d, want 1", v)
	}
	// Closing sends the stream close request on /v3/pipeline.
	if err := db.Close(); err != nil {
		t.Fatal(err)
	}

	stub.mu.Lock()
	defer stub.mu.Unlock()
	if len(stub.requests) < 3 {
		t.Fatalf("stub saw %d requests, want at least 3", len(stub.requests))
	}
	seen := map[string]bool{}
	for _, req := range stub.requests {
		seen[req.path] = true
		if key == "" {
			if req.hasHeader {
				t.Errorf("%s: unexpected %s header %q", req.path, encHeaderName, req.key)
			}
		} else if !req.hasHeader || req.key != key {
			t.Errorf("%s: got %s header %q (present=%v), want %q",
				req.path, encHeaderName, req.key, req.hasHeader, key)
		}
	}
	if !seen["/v3/pipeline"] || !seen["/v3/cursor"] {
		t.Fatalf("both endpoints must be exercised, saw %v", seen)
	}
}
